harness/determined/common/api/_util.py

```diff
@@ -37,14 +38,47 @@ def read_paginated(
     get_with_offset: Callable[[int], T],
     offset: int = 0,
     pages: PageOpts = PageOpts.all,
-) -> Iterator[T]:
+    smart_flatten: bool = False,
+) -> Iterator[Any]:
+    """Read paginated responses from the Determined REST API.
+
+    Args:
+        get_with_offset: A function that takes an offset and returns a bindings.Paginated response.
+        offset: The very first index to start reading from.
+        pages: The number of pages to read. If set to PageOpts.single, only the first page will be
+            read. If set to PageOpts.all, all pages will be read.
+        smart_flatten: An option to yield individual items (instead of pages) from the pages of
+            responses.
+
+    Yields:
+        If smart_flatten is True:
+            individual items from a paginated response
+        If smart_flatten is False:
+            entire response objects (which consiste of a list of individual items and a
+            bindings.v1Pagination for the entire response)
+    """
     while True:
         resp = get_with_offset(offset)
+        # TODO: when on Python 3.8 can use typing.get_args instead of __args__
+        # TODO: when on Python 3.10 isinstance can take a Union type without need of args
+        assert isinstance(
+            resp, bindings.Paginated.__args__  # type: ignore
+        ), "Passed 'get_with_offset' must return a bindings.Paginated type"
         pagination = resp.pagination
         assert pagination is not None
         assert pagination.endIndex is not None
         assert pagination.total is not None
-        yield resp
+        if smart_flatten:
+            resp_attrs = set(resp.__dict__.keys())
+            resp_attrs.remove("pagination")
+            assert (
+                len(resp_attrs) == 1
+            ), "bindings.Paginated responses contain only one attribute besides pagination"
```

Comment:
This could account for those functions that include extra, irrelevant attributes (like model name or something when returning a list of model versions (which themselves have the model name in attributes)) by checking if those attributes are iterable. Hopefully the irrelevant attributes that might be present won't also be iterable.